src/controller.rs

```diff
@@ -46,46 +46,136 @@ struct Context {
     client: Client,
     kubecfg_image: String,
     kubit_image: String,
+    config_map_name: Option<String>,
     only_paused: bool,
 }
 
-fn error_policy(app_instance: Arc<AppInstance>, error: &Error, _ctx: Arc<Context>) -> Action {
+fn error_policy_app_instance(
+    app_instance: Arc<AppInstance>,
+    error: &Error,
+    _ctx: Arc<Context>,
+) -> Action {
     let name = app_instance.name_any();
     warn!(?name, %error, "reconcile failed");
     // TODO(mkm): make error requeue duration configurable
     Action::requeue(Duration::from_secs(5))
 }
 
+fn error_policy_config_map(
+    config_map: Arc<ConfigMap>,
+    error: &Error,
+    _ctx: Arc<Context>,
+) -> Action {
+    let config = &config_map.as_ref().data.as_ref().unwrap()["app-instance"];
+    let ai: Result<AppInstance, serde_yaml::Error> = serde_yaml::from_str(config);
+    match ai {
+        Ok(ai) => error_policy_app_instance(Arc::new(ai), error, _ctx),
+        Err(_) => {
+            error!(%error, "failed to convert configmap to appinstance");
+            std::process::exit(1);
```

Comment:
Bit of a thought here, although might be a wider discussion too. I think this is fine, but putting thoughts out there anyway :smile: 

Since this entirely shuts down on error, I think it could be difficult to catch the logs of "failed to convert...", i.e. the pod crashes and then will instantly restart so you lose the logs. If they go to something like Loki it's okay, since they're captured, but I wonder whether there's a nice way to handle that.

I was initially thinking this could still have the `error!` call, but handle it _somehow_ (hand-wavey). Although in that case the pod would still be `Running`, which could be confusing as to why nothing is happening, only to see a huge `ERROR ...` on a `kubectl logs` call.

---

After writing that/thinking about it a bit more, I think having it crash/restart is actually okay since if you're running in "ConfigMap mode" and the controller keeps crashing, you're more inclined to look at it to see what is wrong, compared to a silent failure.

Maybe there's a nicer way though, perhaps not without external tools though :shrug:

Reply:
yes; while I do agree that there are certain use cases where it's fine to just "let it crash"...

1. as Jack said, the k8s log model is not super-friendly with that model
2. the k8s controller SDKs (and kube-rs in particular) are designed to have a per-resource fine grained error handing (including handling exponential backoffs). Hard crashing at the first error encountered is not good practice in this context